lib/dag/dag.rb

```diff
@@ -1,300 +1,340 @@
-
-    #check that class_name is filled
-    if conf[:link_class_name].nil?
+    options = {
+      class_name: nil,
+      prefix: "",
+      ancestor_class_names: [],
+      descendant_class_names: [],
+      scoped_record_id_column: nil
+    }.merge(options)
+
+    # check that class_name is filled
+    if options[:link_class_name].nil?
       raise ActiveRecord::ActiveRecordError, "has_dag_links must be provided with :link_class_name option"
     end
 
-    #add trailing '_' to prefix
-    unless conf[:prefix] == ''
-      conf[:prefix] += '_'
-    end
-
-    prefix = conf[:prefix]
-    dag_link_class_name = conf[:link_class_name]
-    dag_link_class = conf[:link_class_name].constantize
+    # add trailing '_' to prefix
+    options[:prefix] = "#{options[:prefix]}_" unless options[:prefix] == ""
 
-    if dag_link_class.acts_as_dag_polymorphic?
-      self.class_eval <<-EOL
-              has_many :#{prefix}links_as_ancestor, :as => :ancestor, :class_name => '#{dag_link_class_name}'
-              has_many :#{prefix}links_as_descendant, :as => :descendant, :class_name => '#{dag_link_class_name}'
+    prefix = options[:prefix]
+    dag_link_class_name = options[:link_class_name]
+    dag_link_class = options[:link_class_name].constantize
+    if options[:scoped_record_id_column].present?
+      scoped_record_condition = ",#{options[:scoped_record_id_column]}: record.#{options[:scoped_record_id_column]}"
+      lambda_scoped_record_condition = "lambda { |record| where(#{options[:scoped_record_id_column]}:
+                                       record.#{options[:scoped_record_id_column]}) },".squish
 
-              has_many :#{prefix}links_as_parent, lambda { where('#{dag_link_class.direct_column_name}' => true) }, :as => :ancestor, :class_name => '#{dag_link_class_name}'
-              has_many :#{prefix}links_as_child, lambda { where('#{dag_link_class.direct_column_name}' => true) }, :as => :descendant, :class_name => '#{dag_link_class_name}'
+      class_eval <<-EOL0, __FILE__, __LINE__ + 1
+        attr_accessor :#{options[:scoped_record_id_column]}
+      EOL0
+    else
+      class_eval <<-EOL0, __FILE__, __LINE__ + 1
+        attr_accessor :scoped_record_id
+      EOL0
+    end
 
-      EOL
+    if dag_link_class.acts_as_dag_polymorphic?
+      class_eval <<-EOL1, __FILE__, __LINE__ + 1
+        has_many :#{prefix}links_as_ancestor,#{lambda_scoped_record_condition} as: :ancestor, class_name: "#{dag_link_class_name}"
+        has_many :#{prefix}links_as_descendant,#{lambda_scoped_record_condition} as: :descendant, class_name: "#{dag_link_class_name}"
+        has_many :#{prefix}links_as_parent, lambda { |record| where(#{dag_link_class.direct_column_name}: true#{scoped_record_condition}) }, as: :ancestor, class_name: "#{dag_link_class_name}"
+        has_many :#{prefix}links_as_child, lambda { |record| where(#{dag_link_class.direct_column_name}: true#{scoped_record_condition}) }, as: :descendant, class_name: "#{dag_link_class_name}"
```

Comment:
Tady využívám předpřipravené podmínky, takže pak mám v modelu například
`has_many :links_as_ancestor, lambda { |record| where(demand_workflow_id:              record.demand_workflow_id) }, as: :ancestor, class_name: "DemandWorkflows::WorkflowTree"` a
`has_many :links_as_child, lambda { |record| where(direct: true, demand_workflow_id: record.demand_workflow_id) }, as: :descendant, class_name: "DemandWorkflows::WorkflowTree"`

Používá se to i níže v dalších vazbách.